nvse/nvse/NVSECompiler.cpp

```diff
@@ -0,0 +1,760 @@
+	if (stmt->elseBlock) {
+		// OP_ELSE
+		add_u16(0x17);
+
+		// OP_LEN
+		add_u16(0x02);
+
+		// Placeholder JMP_OPS
+		auto elsePatch = add_u16(0x0);
+
+		// Compile else block
+		set_u16(elsePatch, compileBlock(stmt->elseBlock, true));
+	}
+
+	// OP_ENDIF
+	add_u16(0x19);
+	add_u16(0x0);
+}
+
+void NVSECompiler::visitReturnStmt(const ReturnStmt* stmt) {}
+
+void NVSECompiler::visitWhileStmt(const WhileStmt* stmt) {
+	// OP_WHILE
+	add_u16(0x153B);
+
+	// Placeholder OP_LEN
+	auto exprPatch = add_u16(0x0);
+	auto exprStart = data.size();
+
+	auto jmpPatch = add_u32(0x0);
+
+	// 1 param
+	add_u8(0x1);
+
+	// Compile / patch condition
+	auto condStart = data.size();
+	auto condPatch = add_u16(0x0);
+	stmt->cond->accept(this);
+	set_u16(condPatch, data.size() - condStart);
+
+	// Patch OP_LEN
+	set_u16(exprPatch, data.size() - exprStart);
+
+	// Compile block
+	stmt->block->accept(this);
+
+	// OP_LOOP
+	add_u16(0x153c);
+	add_u16(0x0);
+
+	// Patch jmp
+	set_u32(jmpPatch, data.size());
+}
+
+uint32_t NVSECompiler::compileBlock(StmtPtr& stmt, bool incrementCurrent) {
+	// Store old block size
+	auto temp = result;
+
+	stmt->accept(this);
+
+	auto res = result;
+	if (incrementCurrent) {
+		temp += res;
+	}
+
+	result = temp;
+
+	return res;
+}
+
+void NVSECompiler::visitBlockStmt(const BlockStmt* stmt) {
+	for (auto& statement : stmt->statements) {
+		statement->accept(this);
+	}
+
+	result = stmt->statements.size();
+}
+
+void NVSECompiler::visitAssignmentExpr(const AssignmentExpr* expr) {
+	const auto varInfo = script->varList.GetVariableByName(expr->name.lexeme.c_str());
+	if (!varInfo) {
+		throw std::runtime_error(std::format("Cannot assign to '{}'. Variable with name not found.", expr->name.lexeme));
+	}
+
+	// OP_LET
+	add_u16(0x1539);
+
+	//  Placeholder [expr_len]
+	auto exprPatch = add_u16(0x0);
+	auto exprStart = data.size();
+
+	// Num args
+	add_u8(0x1);
+
+	// Placeholder [param1_len]
+	auto argStart = data.size();
+	auto argPatch = add_u16(0x0);
+
+	// Add arg to stack
+	add_u8('V');
+	add_u8(varInfo->type);
+	add_u16(0); // SCRV
+	add_u16(varInfo->idx); // SCDA
+
+	// Build expression
+	expr->expr->accept(this);
+
+	// OP_ASSIGN
+	add_u8(0);
+
+	// Patch lengths
+	set_u16(argPatch, data.size() - argStart);
+	set_u16(exprPatch, data.size() - exprStart);
+}
+
+void NVSECompiler::visitTernaryExpr(const TernaryExpr* expr) {}
+
+void NVSECompiler::visitBinaryExpr(const BinaryExpr* expr) {
+	expr->left->accept(this);
+	expr->right->accept(this);
+	add_u8(operatorMap[expr->op.lexeme]);
+}
+
+void NVSECompiler::visitUnaryExpr(const UnaryExpr* expr) {
+	expr->expr->accept(this);
+	add_u8(operatorMap[expr->op.lexeme]);
+}
+
+void NVSECompiler::visitCallExpr(const CallExpr* expr) {
+	auto stackRefExpr = dynamic_cast<GetExpr*>(expr->left.get());
+	auto identExpr = dynamic_cast<IdentExpr*>(expr->left.get());
+
+	std::string name{};
+	if (stackRefExpr) {
+		name = stackRefExpr->token.lexeme;
+	} else {
+		name = identExpr->name.lexeme;
```

Comment:
`identExpr` might be null from dynamic cast but not sure if you coded it so it would never happen